server/marketplace/marketplace.go

```diff
@@ -0,0 +1,70 @@
+package marketplace
+
+import (
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+
+	"time"
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+const (
+	// marketplaceVersion is the URL version of Marketplace API.
+	marketplaceVersion = "/api/v1"
+
+	// listPluginsEndpoint is an endpoint to get plugin list.
+	listPluginsEndpoint = "/plugins"
+)
+
+const (
+	// requestTimeout is used as a timeout value to cancel long running request made to Marketplace API.
+	requestTimeout = time.Second * 10
+)
+
+// Marketplace is a gateway to interact with Mattermost Plugin Marketplace.
+type Marketplace struct {
+	// addr of the marketplace.
+	addr string
+
+	// client is used to perform HTTP requests to Marketplace server.
+	client *http.Client
+}
+
+// New creates a new Marketplace with given Marketplace address addr.
+func New(addr string) *Marketplace {
+	return &Marketplace{
+		addr:   addr,
+		client: &http.Client{Timeout: requestTimeout},
+	}
+}
+
+// ListPlugins fetches all plugins from the Marketplace.
+func (m *Marketplace) ListPlugins() (Plugins, error) {
+	urlParsed, err := url.Parse(m.addr)
+	if err != nil {
+		return nil, err
+	}
+	urlParsed.Path = path.Join(marketplaceVersion, listPluginsEndpoint)
+	req, err := http.NewRequest("GET", urlParsed.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := m.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
We can make use of `m.client.Get(urlParsed.String())` directly.

Reply:
done by 556f760477a45a208ac06624deb8d26933b9c13d